lib/qesapdeployment.pm

```diff
@@ -91,6 +102,95 @@ sub qesap_get_deployment_code {
     }
+    assert_script_run("ln -s " . TERRAFORM_DIR . "/aws " . TERRAFORM_DIR . "/ec2");
+}
+
+=head3 qesap_upload_logs
+
+    Collect and upload logs present in @log_files.
+=cut
+sub qesap_upload_logs {
+    record_info("Uploading logfiles", join("\n", @log_files));
+    for my $file (@log_files) {
+        upload_logs($file, failok => 1);
+    }
+    # Remove uploaded files from arrays
+    @log_files = ();
+}
+
+=head3 qesap_yaml_replace
+
+    Replaces yaml config file variables with the ones defined in openqa.
+    Openqa variables need to be added as 'openqa_variables' inside %args as following:
+    <OPENQA_VARIABLE_UPPERCASE> = "<VARIABLE_VALUE>"
+
+=cut
+sub qesap_yaml_replace {
+    my (%run_args) = @_;
+    my %variables = $run_args{openqa_variables};
+    my %replaced_variables = ();
+
+    for my $variable (keys %variables) {
+        $replaced_variables{"%" . $variable . "%"} = $variables{$variable};
+    }
+
+    file_content_replace(QESAP_CONF_TRGT, %replaced_variables);
+}
+
+=head3 qesap_execute
+
+    execute qesap glue script commands
+=cut
+sub qesap_execute {
+    my (%args) = @_;
+    my $verbose = $args{verbose} ? "--verbose" : "";
+    my $qesap_cmd = DEPLOYMENT_DIR . "/scripts/qesap/qesap.py " . $verbose . " -c " . QESAP_CONF_TRGT . " -b " . DEPLOYMENT_DIR;
+    my $command = defined($args{cmd}) ? $args{cmd} : die("QESAP command to execute undefined");
+    my $exec_log = "/root/qesap_exec_" . $command . ".log";
+
+    record_info('QESAP exec', 'Executing: \n' . $qesap_cmd . " " . $command);
+    assert_script_run($qesap_cmd . " " . $command . " > " . $exec_log, quiet => 1);
+    push(@log_files, $exec_log);
+    qesap_upload_logs();
+}
+
+=head3 qesap_prepare_env
+
+    Prepare terraform environment.
+    - creates file structures
+    - pulls git repository
+    - external config files
+    - installs pip requirements and OS packages
+    - generates config files with qesap script
+
+=cut
+sub qesap_prepare_env {
+    my $pip_cmd = "pip install --no-color --no-cache-dir --progress-bar off ";
+    my $tfvars_template = get_var('QESAP_TFVARS_TEMPLATE');
+    my $curl = "curl -v -L ";
+
+    qesap_get_deployment_code();
+
+    record_info("QESAP repo", "Installing pip requirements");
+    assert_script_run($pip_cmd . " -r " . DEPLOYMENT_DIR . "/requirements.txt > /root/pip_requirements.log");
+    push(@log_files, "/root/pip_requirements.log");
+
+    # Copy tfvars template file if defined in parameters
+    if (get_var('QESAP_TFVARS_TEMPLATE')) {
+        record_info("QESAP tfvars template", "Preparing terraform template: \n" . $tfvars_template);
+        assert_script_run('cd ' . TERRAFORM_DIR, quiet => 1);
+        assert_script_run('cp ' . $tfvars_template . ' terraform.tfvars.template');
+    }
+
+    record_info("QESAP yaml", "Preparing yaml config file");
+    assert_script_run($curl . data_url(QESAP_CONF_SRC) . ' -o ' . QESAP_CONF_TRGT);
+    qesap_yaml_replace();
```

Comment:
If this working? I thought `qesap_yaml_replace()` received at least one argument.

Reply:
Yep. Function is taking the hash of variables from $run_args{openqa_variables} so it is up to the test code to take care of it. But thinking about it now... there is no failure defined if the hash is empty or not defined at all.